### PR TITLE
Add RawHID support to ATSAM (Massdrop boards)

### PR DESCRIPTION
**Pull request by foxx1337**
## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or ~~update~~)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Added 300 ms timeout for the led indicators for Massdrop CTRL
* Added raw HID support together with a very stupid test program linked in the readme. My plan is to write a proper configurator a la VIA, but this is a good starting point as CTRL (and ALT) lacked this arm_atsam functionality.
    The code is originally authored by @helluvamatt and I saw that @ash0x0 also has interest in these things.

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by foxx1337** [comment]
Good point. I didn't consider the work to be core related, but you're right.

**Comment by drashna** [comment]
What zvecr said. This is really cool, and the work is absolutely appreciated!  But if this ca nbe broken out into 2 PRs, one with the core changes and one with the keymap, that would be perfect.   

And glad to see raw HID support added.

**Comment by foxx1337** [comment]
Do you think it makes sense to actually remove the keymap itself and include the code in the default?

It should be a normal thing to have the leds turn off after a while (5 minutes now). Also the code I added to "my" keymap helps showcase the "core" change itself. What do you think, @drashna, @zvecr ?

**Comment by zvecr** [comment]
Given RawHID isnt a new feature, and this PR aims to fill in missing compatibility, there is no really need to demo that feature in a keymap.

**Comment by drashna** [comment]
Exactly. A PR to add the core functionality for ATSAM, and a PR to add your keymap, leveraging the functionality. 

Specifically, from the contributing docs:
https://docs.qmk.fm/#/contributing?id=quantumtmk-core
> Do not lump keyboards or keymaps in with core changes. Submit your core changes first.

**Comment by foxx1337** [comment]
I will remove my keymap since it has nothing of value then. Be back later with the changes and much appreciating your work, guys.

**Comment by foxx1337** [comment]
Note that the title doesn't apply anymore at this point. No LED timeout there.

**Comment by drashna** [comment]
I wouldn't say that it had nothing of value, at all!  At the very least, it was a great example of the type of stuff that Raw HID can do.  

But thank you!  

Also, I think you can edit the title. If not....

**Comment by tzarc** [comment]
No issues from my end, just leaving it open for @zvecr as he was specifically requested earlier.

File: tmk_core/protocol/arm_atsam/main_arm_atsam.c [hunk 1]
```diff
@@ -206,10 +206,20 @@ void main_subtask_usb_extra_device(void) {
...
 
-void main_subtasks(void) {
+#ifdef RAW_ENABLE
+void main_subtask_raw(void) {
+    udi_hid_raw_receive_report();
+}
+#endif
+
+void main_subtasks(void)
+{
```
Comment (fauxpark) [inline tmk_core/protocol/arm_atsam/main_arm_atsam.c]:
```suggestion
void main_subtasks(void) {
```

File: tmk_core/protocol/arm_atsam/usb/main_usb.c [hunk 2]
```diff
@@ -88,5 +91,9 @@ bool          main_raw_enable(void) {
     return true;
 }
 
-void main_raw_disable(void) { main_b_raw_enable = false; }
+void          main_raw_disable(void) { main_b_raw_enable = false; }
+
+void          main_raw_receive(uint8_t *buffer, uint8_t len) {
```
Comment (fauxpark) [inline tmk_core/protocol/arm_atsam/usb/main_usb.c]:
```suggestion
void main_raw_disable(void) { main_b_raw_enable = false; }

void main_raw_receive(uint8_t *buffer, uint8_t len) {
```

File: tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c [hunk 3]
```diff
@@ -719,7 +723,31 @@ static void udi_hid_raw_report_sent(udd_ep_status_t status, iram_size_t nb_sent,
...
+bool udi_hid_raw_receive_report(void) {
+    if (!main_b_raw_enable) {
+        return false;
+    }
+
+    return udd_ep_run(UDI_HID_RAW_EP_OUT | USB_EP_DIR_OUT, false, udi_hid_raw_report_recv, UDI_HID_RAW_REPORT_SIZE, udi_hid_raw_report_rcvd);
+}
+
+static void udi_hid_raw_report_rcvd(udd_ep_status_t status, iram_size_t nb_rcvd, udd_ep_id_t ep)
+{
```
Comment (fauxpark) [inline tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c]:
```suggestion
static void udi_hid_raw_report_rcvd(udd_ep_status_t status, iram_size_t nb_rcvd, udd_ep_id_t ep) {
```

File: tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c [hunk 4]
```diff
@@ -719,7 +723,31 @@ static void udi_hid_raw_report_sent(udd_ep_status_t status, iram_size_t nb_sent,
...
+      UNUSED(ep);
```
Comment (fauxpark) [inline tmk_core/protocol/arm_atsam/usb/udi_hid_kbd.c]:
```suggestion
    UNUSED(ep);
```

**Comment by foxx1337** [comment]
@drashna I'll return in the next hours with an improved protocol and a configurator program to deal with the raw hid stuff. I didn't just dig for raw hid because of sports, but because I need a configurator :)

**Comment by zvecr** [comment]
Thanks!

**Comment by zvecr** [comment]
Also as a FYI, if you cherry pick #6068 (and the linked mdloader changes) you could just implement a via keymap and not have to create your own configuration tool.

**Pull request merged**